ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java

```diff
@@ -88,226 +80,257 @@ public UdfFactory getUdfFactory(final String functionName) {
-    if (!functionNameValidator.test(factory.getName())) {
-      throw new KsqlException(factory.getName() + " is not a valid function name."
-          + " Function names must be valid java identifiers and not a KSQL reserved word"
-      );
+  public void ensureFunctionFactory(final UdfFactory factory) {
+    validateFunctionName(factory.getName());
+
+    final String functionName = factory.getName().toUpperCase();
+    if (udafs.containsKey(functionName)) {
+      throw new KsqlException("UdfFactory already registered as aggregate: " + functionName);
+    }
+
+    final UdfFactory existing = udfs.putIfAbsent(functionName, factory);
+    if (existing != null && !existing.matches(factory)) {
+      throw new KsqlException("UdfFactory not compatible with existing factory."
+          + " existing: " + existing
+          + ", factory: " + factory);
     }
-    return ksqlFunctionMap.putIfAbsent(factory.getName().toUpperCase(), factory) == null;
   }
 
   @Override
   public boolean isAggregate(final String functionName) {
-    return aggregateFunctionMap.containsKey(functionName.toUpperCase());
+    return udafs.containsKey(functionName.toUpperCase());
   }
 
   @Override
   public KsqlAggregateFunction getAggregate(
       final String functionName,
       final Schema argumentType
   ) {
-    final AggregateFunctionFactory aggregateFunctionFactory
-        = aggregateFunctionMap.get(functionName.toUpperCase());
-    if (aggregateFunctionFactory == null) {
+    final AggregateFunctionFactory udafFactory = udafs.get(functionName.toUpperCase());
+    if (udafFactory == null) {
       throw new KsqlException("No aggregate function with name " + functionName + " exists!");
     }
-    return aggregateFunctionFactory.getProperAggregateFunction(
-        Collections.singletonList(argumentType));
+
+    return udafFactory.getProperAggregateFunction(Collections.singletonList(argumentType));
   }
 
   @Override
   public void addAggregateFunctionFactory(final AggregateFunctionFactory aggregateFunctionFactory) {
-    aggregateFunctionMap.putIfAbsent(
-        aggregateFunctionFactory.getName().toUpperCase(),
-        aggregateFunctionFactory);
-  }
+    final String functionName = aggregateFunctionFactory.getName().toUpperCase();
+    validateFunctionName(functionName);
 
-  @Override
-  public FunctionRegistry copy() {
-    return new InternalFunctionRegistry(
-        new HashMap<>(ksqlFunctionMap),
-        new HashMap<>(aggregateFunctionMap));
+    if (udfs.containsKey(functionName)) {
+      throw new KsqlException(
+          "Aggregate function already registered as non-aggregate: " + functionName);
+    }
+
+    if (udafs.putIfAbsent(functionName, aggregateFunctionFactory) != null) {
+      throw new KsqlException("Aggregate function already registered: " + functionName);
+    }
   }
 
   @Override
   public List<UdfFactory> listFunctions() {
-    return new ArrayList<>(ksqlFunctionMap.values());
+    return new ArrayList<>(udfs.values());
```

Comment:
Not code I've changed.... but I'd noticed this too.

`udfs.values()` is a Collection, not a List. So this would require changing the return value. Let me take a look...